pkg/executor/join/semi_join_probe.go

```diff
@@ -0,0 +1,332 @@
+	if err != nil {
+		return err
+	}
+
+	if s.unFinishedProbeRowIdxQueue.IsEmpty() {
+		// To avoid `Previous chunk is not probed yet` error
+		s.currentProbeRow = s.chunkRows
+	}
+
+	meta := s.ctx.hashTableMeta
+	if joinedChk.NumRows() > 0 {
+		s.selected, err = expression.VectorizedFilter(s.ctx.SessCtx.GetExprCtx().GetEvalCtx(), s.ctx.SessCtx.GetSessionVars().EnableVectorizedExpression, s.ctx.OtherCondition, chunk.NewIterator4Chunk(joinedChk), s.selected)
+		if err != nil {
+			return err
+		}
+
+		for index, result := range s.selected {
+			if result {
+				meta.setUsedFlag(*(*unsafe.Pointer)(unsafe.Pointer(&s.rowIndexInfos[index].buildRowStart)))
+			}
+		}
+	}
+
+	return
+}
+
+func (s *semiJoinProbe) probeForLeftSideBuildNoOtherCondition(sqlKiller *sqlkiller.SQLKiller) (err error) {
+	meta := s.ctx.hashTableMeta
+	tagHelper := s.ctx.hashTableContext.tagHelper
+
+	loopCnt := 0
+
+	for s.currentProbeRow < s.chunkRows {
+		if s.matchedRowsHeaders[s.currentProbeRow] != 0 {
+			candidateRow := tagHelper.toUnsafePointer(s.matchedRowsHeaders[s.currentProbeRow])
+			if !meta.isCurrentRowUsedWithAtomic(candidateRow) && isKeyMatched(meta.keyMode, s.serializedKeys[s.currentProbeRow], candidateRow, meta) {
+				meta.setUsedFlag(candidateRow)
+			} else {
+				s.probeCollision++
+			}
+			s.matchedRowsHeaders[s.currentProbeRow] = getNextRowAddress(candidateRow, tagHelper, s.matchedRowsHashValue[s.currentProbeRow])
+		} else {
+			s.currentProbeRow++
+		}
+
+		loopCnt++
+		if loopCnt%2000 == 0 {
+			err = checkSQLKiller(sqlKiller, "killedDuringProbe")
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	err = checkSQLKiller(sqlKiller, "killedDuringProbe")
+	if err != nil {
+		return err
+	}
+
+	return
+}
+
+func (s *semiJoinProbe) probeForRightSideBuildHasOtherCondition(chk, joinedChk *chunk.Chunk, remainCap int, sqlKiller *sqlkiller.SQLKiller) (err error) {
+	err = s.concatenateProbeAndBuildRows(joinedChk, sqlKiller, true)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if s.unFinishedProbeRowIdxQueue.IsEmpty() && s.offset == s.chunkRows {
+			// To avoid `Previous chunk is not probed yet` error
+			s.currentProbeRow = s.chunkRows
+		}
+	}()
+
+	if joinedChk.NumRows() > 0 {
+		s.selected = s.selected[:0]
+		s.selected, err = expression.VectorizedFilter(s.ctx.SessCtx.GetExprCtx().GetEvalCtx(), s.ctx.SessCtx.GetSessionVars().EnableVectorizedExpression, s.ctx.OtherCondition, chunk.NewIterator4Chunk(joinedChk), s.selected)
+		if err != nil {
+			return err
+		}
+
+		s.setIsMatchedRows()
+	}
+
+	if s.unFinishedProbeRowIdxQueue.IsEmpty() {
+		for remainCap > 0 && (s.offset < s.chunkRows) {
+			rowNumToAppend := int(math.Min(float64(remainCap), float64(s.chunkRows-s.offset)))
+
+			for index, usedColIdx := range s.lUsed {
+				dstCol := chk.Column(index)
+				srcCol := s.currentChunk.Column(usedColIdx)
+				chunk.CopySelectedRowsWithRowIDFunc(dstCol, srcCol, s.isMatchedRows, s.offset, s.offset+rowNumToAppend, func(i int) int {
+					return i
```

Comment:
I think here should return j.usedRows[i], do you have any test case to cover the case of input chunk has `sel` array?